kv/memberlist/memberlist_client.go

```diff
@@ -533,43 +529,83 @@ func (m *KV) JoinMembers(members []string) (int, error) {
 	return m.memberlist.Join(members)
 }
 
-func (m *KV) joinMembersOnStartup(ctx context.Context, members []string) error {
-	reached, err := m.memberlist.Join(members)
-	if err == nil {
-		level.Info(m.logger).Log("msg", "joined memberlist cluster", "reached_nodes", reached)
-		return nil
+// fastJoinMembersOnStartup attempts to reach small subset of nodes (computed as RetransmitMult * log10(number of discovered members + 1)).
+func (m *KV) fastJoinMembersOnStartup(ctx context.Context) {
+	startTime := time.Now()
+
+	members := m.discoverMembers(ctx, m.cfg.JoinMembers)
+	toReach := m.cfg.RetransmitMult * int(math.Ceil(math.Log10(float64(len(members)+1))))
+	if toReach > len(members) {
+		toReach = len(members)
 	}
 
-	if m.cfg.MaxJoinRetries <= 0 {
-		return err
+	level.Info(m.logger).Log("msg", "memberlist fast-join starting", "members_found", len(members), "to_reach", toReach)
+
+	totalReached := 0
+	previouslyAttemptedMembers := map[string]bool{}
+
+	for ; toReach > 0; members = m.discoverMembers(ctx, m.cfg.JoinMembers) {
+		// Remove members we have tried before.
+		for ix := len(members) - 1; ix >= 0; ix-- {
+			if previouslyAttemptedMembers[members[ix]] {
+				members = append(members[:ix], members[ix+1:]...)
+			}
+		}
+
+		// If there are no more nodes to try, we give up.
+		if len(members) == 0 {
+			break
+		}
+
+		if len(members) > toReach {
+			members = members[:toReach]
+		}
+
+		for _, m := range members {
+			previouslyAttemptedMembers[m] = true
+		}
+
+		reached, err := m.memberlist.Join(members)
+		if err == nil {
+			totalReached += reached
+			toReach -= reached
+		} else {
+			level.Debug(m.logger).Log("msg", "failed to reach any nodes while fast-joining memberlist cluster", "attempted_nodes", len(members), "err", err)
+		}
 	}
 
-	level.Debug(m.logger).Log("msg", "attempt to join memberlist cluster failed", "retries", 0, "err", err)
-	lastErr := err
+	level.Info(m.logger).Log("msg", "fast-join memberlist cluster finished", "reached_nodes", totalReached, "elapsed_time", time.Since(startTime))
+}
+
+func (m *KV) joinMembersOnStartup(ctx context.Context) bool {
+	startTime := time.Now()
 
 	cfg := backoff.Config{
 		MinBackoff: m.cfg.MinJoinBackoff,
 		MaxBackoff: m.cfg.MaxJoinBackoff,
 		MaxRetries: m.cfg.MaxJoinRetries,
 	}
 
-	backoff := backoff.New(ctx, cfg)
+	boff := backoff.New(ctx, cfg)
+	var lastErr error
 
-	for backoff.Ongoing() {
-		backoff.Wait()
+	for boff.Ongoing() {
+		members := m.discoverMembers(ctx, m.cfg.JoinMembers)
 
-		reached, err := m.memberlist.Join(members)
-		if err != nil {
-			lastErr = err
-			level.Debug(m.logger).Log("msg", "attempt to join memberlist cluster failed", "retries", backoff.NumRetries(), "err", err)
-			continue
+		reached, err := m.memberlist.Join(members) // err is only returned if reached==0.
```

Comment:
If we have 100 nodes, but only 1 fail, we redo the pull-push against all 100. I think this is inefficient. Since the `Join()` internally iterates over the members, why not just calling it 1-by-1 so that we can known which member succeeded and which failed, and then we re-try only the failed ones?

Reply:
> If we have 100 nodes, but only 1 fail, we redo the pull-push against all 100.

If only 1 node fails, we don't get error. We only get error if `m.memberlist.Join(members)` failed to join ALL nodes.

Reply:
Right!